jayden 인증 과제 step3입니다.

Tarakyu opened the pull request:
## 구현사항
- 관리자를 구현하고, 테마와 스케쥴의 생성과 삭제는 관리자만 가능하도록 제한하였습니다.
- 지난 리뷰에서 말씀해주신 어노테이션 기반의 처리를 더 확장해서 적용해보았습니다.

## 느낀점
- Interceptor가 MemberDao에 의존하는 것이 굉장히 어색합니다. 이런 문제를 해결하는 법을 이번주 과제에서 배운다고 하니 다음에 적용해보도록 하겠습니다.
- 이렇게 모든 플로우를 직접 테스트하는 E2E테스트가 많아지는 경우 기획이 바뀌게 되면 테스트를 고치는데 엄청난 시간이 소요될 것 같습니다. 더 좋은 테스트 방법론도 생각해보아야 할 것 같습니다.

kakao-jayden-v committed:
feat: 멤버 권한 추가

kakao-jayden-v committed:
chore: 초기 데이터 자동생성 구현

kakao-jayden-v committed:
test: 어드민 관련 테스트 코드 수정

- 테마, 스케쥴 생성 및 삭제 테스트게 어드민 권한 추가
- 테마, 스케쥴 생성 및 삭제시 어드민 권한 없는 경우에 대한 테스트 추가

gogyzzz commented:
지난 리뷰 때 제가 말씀을 못드렸는데요, AuthController 에서 토큰을 만들 때 username으로 만들기보다는 민감하지 않은 정보로 만드는게 좋을 것 같아요~ 예를 들어 id 라든지 말이죠 

```java
    @PostMapping("/token")
    public ResponseEntity login(@RequestBody TokenRequest request) {
        Member member = memberService.findByUsername(request.getUsername());
        if (member.checkWrongPassword(request.getPassword())) {
            throw new AuthorizationException("비밀번호가 올바르지 않습니다.");
        }
        String accessToken = jwtTokenProvider.createToken(request.getUsername());
        return ResponseEntity.ok().body(new TokenResponse(accessToken));
    }
```

gogyzzz commented on src/main/java/nextstep/auth/LoginInterceptor.java:
```diff
@@ -1,35 +1,51 @@
+            Member member = memberDao.findByUsername(username)
+                    .orElseThrow(NotExistEntityException::new);
+            if (!member.isAdmin()) {
+                throw new AuthorizationException("접근 권한이 없습니다.");
```
다른 분 리뷰 보다가 알게 된 건데요, AuthorizationException은 인증 과정에서 에러가 나는 경우에 한해 쓰더라구요. (토큰이 잘못됐다거나...)

그래서 이런 경우는 Forbidden 에러를 쓰도록 바꿔보는게 어떨까 싶습니다~
(token은 유효한데(등록된 누군가는 맞는데) 특정 권한이 없는 경우)

Advice에서도 AuthorizationException 과 ForbiddenException을 나누고 각각 401, 403으로 처리하도록이요!

gogyzzz commented:
ReservationDao의 findByScheduleId 메서드에서 예외가 발생할 경우 null 리턴을 하는데요, 이 메서드를 사용하는 ReservationService::create 에서 List로 다루고 있는 것 같아요.
템플릿 코드이긴 한데 nullable로 받거나 null을 리턴하지 않도록 변경해야 할 것 같습니다~ 

```java
// ReservationDao
    public List<Reservation> findByScheduleId(Long id) {
        String sql = "SELECT reservation.id, reservation.schedule_id, reservation.name, schedule.id, schedule.theme_id, schedule.date, schedule.time, theme.id, theme.name, theme.desc, theme.price " +
                "from reservation " +
                "inner join schedule on reservation.schedule_id = schedule.id " +
                "inner join theme on schedule.theme_id = theme.id " +
                "where schedule.id = ?;";

        try {
            return jdbcTemplate.query(sql, rowMapper, id);
        } catch (Exception e) {
            return null;
        }
// ReservationService
        List<Reservation> reservation = reservationDao.findByScheduleId(schedule.getId());
        if (!reservation.isEmpty()) {
            throw new DuplicateEntityException();
        }

```

gogyzzz commented:
AuthorizationExtractor 의 extract 에서도 null 리턴을 하는데요, 이걸 사용하는 부분에서는 전부 String으로 받고 있더라구요~ 이 부분도 Optional 등으로 처리하는게 좋겠습니다

```java
// AuthorizationExtractor
    public static String extract(HttpServletRequest request) {
        Enumeration<String> headers = request.getHeaders(AUTHORIZATION);
        while (headers.hasMoreElements()) {
            String value = headers.nextElement();
            if ((value.toLowerCase().startsWith(BEARER_TYPE.toLowerCase()))) {
                String authHeaderValue = value.substring(BEARER_TYPE.length()).trim();
                request.setAttribute(ACCESS_TOKEN_TYPE, value.substring(0, BEARER_TYPE.length()).trim());
                int commaIndex = authHeaderValue.indexOf(',');
                if (commaIndex > 0) {
                    authHeaderValue = authHeaderValue.substring(0, commaIndex);
                }
                return authHeaderValue;
            }
        }

        return null;
    }
```

gogyzzz commented:
고생하셨습니다~ 주말 잘 보내세요!

kakao-jayden-v committed:
feat: JWT에 username이 아닌 id를 저장하도록 변경

kakao-jayden-v committed:
feat: 접근 권한이 없는 경우 403 에러 반환하도록 변경

kakao-jayden-v committed:
fix: null 예외처리

- Database 오류 발생시 null을 반환하지 않고 에러를 반환하도록 변경
- AuthorizationExtractor의 extract 메서드가 Optional을 반환하도록 변경

Tarakyu commented:
JWT의 내용은 모두가 확인할 수 있다는 점을 놓치고 있었네요!! 바로 수정하였습니다 🥲
인증과 인가의 차이에 대해서도 다시 공부해보았습니다.
템플릿 코드의 null 처리 방식도 개선해보았습니다! `return null`은 너무 끔찍한 것 같아요
